## Snakefile for analyses into the imputation and phasing of low cov sequencing data in Anopheles gambiae
import pandas as pd
import numpy as np
configfile: "config/config.yaml"

#chroms = ['2L', '2R', '3L', '3R', 'X']
chroms = ['2L', 'X']

metadata = pd.read_csv("resources/samples.meta.txt", sep="\t")

lowCovMeta = pd.read_csv("config/GhanaLowCov.metadata.tsv", sep="\t")
bool_ = lowCovMeta['Site'].isin(['Asokwa', 'Bantama-Kumasi', 'Manso-Nkwanta'])
samples = lowCovMeta['sample'][~bool_][:10]

# IR allele balance mutations file
mutationData = pd.read_csv("resources/coemutations.tsv", sep="\t")

rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
       # expand("resources/ag1000g_WestAfrica_col_{chrom}.vcf.gz", chrom=chroms),
        expand("results/alignments/{sample}.bam.bai", sample=samples),
        expand("results/alignments/coverage/{sample}.mosdepth.summary.txt", sample=samples),
        #expand("results/vcfs/{sample}.calls.{chrom}.vcf.gz.csi", sample=samples, chrom=chroms),
        #expand("results/vcfs/merged_calls.{chrom}.vcf.gz", chrom=chroms),


include: "rules/prepHapPanel.smk"
include: "rules/prepLowCovSamples.smk"
include: "rules/impute.smk"